pennylane/_qubit_device.py

```diff
@@ -0,0 +1,282 @@
+
+        Args:
+            queue (Iterable[~.operation.Operation]): operations to execute on the device
+            observables (Iterable[~.operation.Observable]): observables to measure and return
+            parameters (dict[int->list[ParameterDependency]]): Mapping from free parameter index to the list of
+                :class:`Operations <pennylane.operation.Operation>` (in the queue) that depend on it.
+
+        Raises:
+            QuantumFunctionError: if the value of :attr:`~.Observable.return_type` is not supported
+
+        Returns:
+            array[float]: measured value(s)
+        """
+        if parameters is None:
+            parameters = {}
+
+        self.check_validity(queue, observables)
+        self._op_queue = queue
+        self._obs_queue = observables
+        self._parameters = {}
+        self._parameters.update(parameters)
+
+        results = []
+
+        with self.execution_context():
+            self.pre_apply()
+
+            for operation in queue:
+                # Pass instances directly
+                self.apply(operation)
+
+            self.post_apply()
+
+            self.pre_measure()
+
+            for obs in observables:
+                # Pass instances directly
+                if obs.return_type is Expectation:
+                    results.append(self.expval(obs))
+
+                elif obs.return_type is Variance:
+                    results.append(self.var(obs))
+
+                elif obs.return_type is Sample:
+                    results.append(np.array(self.sample(obs)))
+
+                elif obs.return_type is Probability:
+                    results.append(list(self.probability(wires=obs.wires).values()))
+
+                elif obs.return_type is not None:
+                    raise QuantumFunctionError("Unsupported return type specified for observable {}".format(obs.name))
+
+            self.post_measure()
+
+            self._op_queue = None
+            self._obs_queue = None
+            self._parameters = None
+
+            # Ensures that a combination with sample does not put
+            # expvals and vars in superfluous arrays
+            if all(obs.return_type is Sample for obs in observables):
+                return self._asarray(results)
+            if any(obs.return_type is Sample for obs in observables):
+                return self._asarray(results, dtype="object")
+
+            return self._asarray(results)
+
+    def rotate_basis(self, observable):
+        """Rotates the specified wires such that they
+        are in the eigenbasis of the provided observable.
+
+        Args:
+            observable (Observable): the observable that is to be
+                measured
+        """
+        for diag_gate in observable.diagonalizing_gates():
+            self.apply(diag_gate)
+
+    def expval(self, observable):
+        wires = observable.wires
+        self.rotate_basis(observable)
```

Comment:
I wonder if this is the right spot for this. Do you think one could move this to a separate method happening before the measurement?

Reply:
Or maybe in the default implementation of `pre_measure`?

Reply:
Yes, it was considered to be moved to ``pre_measure``, although then simply calling ``dev.expval`` or any other measurement function will not include a call on ``dev.rotate_basis()``. This would make porting the existing tests more troublesome, as they all assume that ``dev.queue`` already includes all operations from the circuit.

Reply:
After more exploration, it has been decided to move this into a global location (like `pre_measure`) in the following PR.